# Codebook  
**Authors:** Patrick Guo  
Documenting existing data files of DaanMatch with information about location, owner, "version", source etc.

In [1]:
import boto3
import numpy as np 
import pandas as pd
pd.plotting.register_matplotlib_converters()
import matplotlib.pyplot as plt
%matplotlib inline
from collections import Counter
import statistics

In [2]:
client = boto3.client('s3')
resource = boto3.resource('s3')
my_bucket = resource.Bucket('my-bucket')

# 42621 Final_Data_ngodarpan.gov.in

## TOC:
* [About this dataset](#1)
* [What's in this dataset](#2)
* [Codebook](#3)
    * [Missing values](#3.1)
    * [Summary statistics](#3.2)
* [Columns](#4)
    * [Name](4.1)
    * [ngo url](4.2)
    * [Mobile](4.3)
    * [UniqueID](4.4)
    * [Off phone1](4.5)
    * [Email](4.6)
    * [Major Activities1](4.7)
    * [operational states db](4.8)
    * [issues working db](4.9)
    * [operational district db](4.10)
    * [reg name](4.11)
    * [fcrano](4.12)
    * [nr regNo](4.13)
    * [nr add](4.14)
    * [nr orgName](4.15)
    * [ngo reg date](4.16)
    * [nr actName](4.17)
    * [nr city](4.18)
    * [TypeDescription](4.19)
    * [StateName](4.20)
    * [status](4.21)
    * [president name](4.22)
    * [president email](4.23)
    * [president mobile](4.24)
    * [Chairman name](4.25)
    * [Chairman email](4.26)
    * [Chairman mobile](4.27)
    * [Secretary name](4.28)
    * [Secretary email](4.29)
    * [Secretary mobile](4.30)
    * [Asisstant Secretary name](4.31)
    * [Asisstant Secretary email](4.32)
    * [Asisstant Secretary mobile](4.33)
    * [Board Member name](4.34)
    * [Board Member email](4.35)
    * [Board Member mobile](4.36)
    * [Vice Chairman name](4.37)
    * [Vice Chairman email](4.38)
    * [Vice Chairman mobile](4.39)
    * [Member name](4.40)
    * [Member email](4.41)
    * [Member mobile](4.42)

In [74]:
# Lists out the column names in TOC format
def toc_maker(dataset):
    counter = 1
    for column in dataset.columns:
        print("* ["+column+"](4."+str(counter)+")")
        counter +=1

In [1]:
#toc_maker(Final_Data_ngodarpan)

**About this dataset**  <a class="anchor" id="1"></a>  
Data provided by: NGO Darpan  
Source: ngodarpan.gov.in   
Type: xlsx  
Last Modified: June 1, 2021, 17:06:30 (UTC-07:00)  
Size: 49.7 MB

In [7]:
path = "s3://daanmatchdatafiles/Darpan21FCRA/42621 Final_Data_ngodarpan.gov.in.xlsx"
xl = pd.ExcelFile(path)
print(xl.sheet_names)
Final_Data_ngodarpan = xl.parse('ngodarpan.gov.in')
Final_Data_ngodarpan.head()

['ngodarpan.gov.in']


Name                    ngo url        Mobile  \
0                             PRAYAS                        NaN  9.778080e+09   
1        PONDICHERRYWOMENSCONFERENCE                        NaN  9.443253e+09   
2           SHABRI SAMAJ SEWA SAMITI  http://ssssamitibhind.org  7.828394e+09   
3  ANAND GANGA SAMAJIK SIKSHA SAMITI                        NaN  9.450678e+09   
4        Himaliyan Gram Vikas Samiti                        NaN  9.412037e+09   

          UniqueID    Off phone1                        Email  \
0  OR/2009/0010000  06858-223440    director_prayas@yahoo.com   
1  PY/2016/0100001  0413-2213238           surebe33@gmail.com   
2  MP/2016/0100003  0751-1234689     ssssamitibhind@gmail.com   
3  UP/2016/0100004  05566-281059  lovelyraivijendra@gmail.com   
4  UA/2016/0100009  05964-213271           hgvs1990@gmail.com   

                                   Major Activities1 operational states db  \
0                           1.63 Nos. of SHGs formed               ORISSA,   
1  Working for Women and Children Obtaining Loan ...           PUDUCHERRY,   
2  more than one thousand leadership development ...       MADHYA PRADESH,   
3  OUR ORGANISATION HAVE PLANTED MORE THAN 2 LAKH...        UTTAR PRADESH,   
4  Facilitated formation and strengthening of 65C...          UTTARAKHAND,   

                                   issues working db  \
0  Agriculture,Children,Civic Issues,Disaster Man...   
1        Women's Development & Empowerment,Children,   
2  Animal Husbandry, Dairying & Fisheries,Agricul...   
3  Agriculture,Environment & Forests,Health & Fam...   
4  Animal Husbandry, Dairying & Fisheries,Agricul...   

                             operational district db  ...  \
0                             ORISSA->Nabarangapur ,  ...   
1                            PUDUCHERRY->Puducherry,  ...   
2  MADHYA PRADESH->Anuppur, Ashoknagar, Balaghat,...  ...   
3  UTTAR PRADESH->Deoria, Gorakhpur, Sant Kabir N...  ...   
4  UTTARAKHAND->Almora , Bageshwar, Champawat, Pi...  ...   

  Asisstant Secretary mobile Board Member name   Board Member email  \
0                        NaN               NaN                  NaN   
1                        NaN               NaN                  NaN   
2                        NaN               NaN                  NaN   
3                        NaN               NaN                  NaN   
4                        NaN      Krishna Nand  hgvsgan@yahoo.co.in   

  Board Member mobile Vice Chairman name           Vice Chairman email  \
0                 NaN                NaN                           NaN   
1                 NaN                NaN                           NaN   
2                 NaN                NaN                           NaN   
3                 NaN                NaN                           NaN   
4        7.500720e+09   Leela Dhar Joshi  hgvs.jleeladhar.lj@gmail.com   

  Vice Chairman mobile Member name              Member email Member mobile  
0                  NaN         NaN                       NaN           NaN  
1                  NaN         NaN                       NaN           NaN  
2                  NaN       ALOK   ssssamitibhind@gmail.com  7.828498e+09  
3                  NaN         NaN                       NaN           NaN  
4         8.057816e+09         NaN                       NaN           NaN  

[5 rows x 42 columns]

**What's in this dataset?**  <a class="anchor" id="2"></a>  

In [8]:
dataset = Final_Data_ngodarpan
print("Shape:", dataset.shape)
print("Rows:", dataset.shape[0])
print("Columns:", dataset.shape[1])
print("Each row is a NGO.")

Shape: (111929, 42)
Rows: 111929
Columns: 42
Each row is a NGO.


**Codebook** <a class="anchor" id="3"></a>

In [2]:
dataset_columns = [column for column in dataset.columns]
dataset_desc = ["Name of NGO",
               "URL for NGO",
               "Mobile phone",
               "Unique ID of VO/NGO",
               "Telephone/Alternate number",
               "Email address",
               "Description of major activities",
               "List of states they operate in",
               "List of issues they are working on",
               "List of districts they operate in",
               "Name of registrar",
               "FCRA number",
               "Registration number",
               "Address",
               "Name of NGO",
               "Registration date",
               "Name of Act",
               "City of NGO",
               ]
dataset_desc = dataset_desc + ["N/A"] * (len(dataset_columns) - len(dataset_desc))
dataset_dtypes = [dtype for dtype in dataset.dtypes]

data = {"Column Name": dataset_columns, "Description": dataset_desc, "Type": dataset_dtypes}
dataset_codebook = pd.DataFrame(data)
dataset_codebook

NameError: name 'dataset' is not defined

**Missing values** <a class="anchor" id="3.1"></a>

In [10]:
Final_Data_ngodarpan.isnull().sum()

Name                               0
ngo url                        86142
Mobile                            32
UniqueID                           0
Off phone1                     95402
Email                              0
Major Activities1              27311
operational states db          23039
issues working db              22637
operational district db        23039
reg name                           0
fcrano                         89869
nr regNo                           3
nr add                             0
nr orgName                         0
ngo reg date                       0
nr actName                      1316
nr city                          214
TypeDescription                    0
StateName                          0
status                        111929
president name                 52520
president email                52520
president mobile               52520
Chairman name                  82126
Chairman email                 82132
Chairman mobile                82137
S

**Summary statistics** <a class="anchor" id="3.2"></a>

None. All qualitative features.

## Columns
<a class="anchor" id="4"></a>

### Name
<a class="anchor" id="4.1"></a>
Name of NGO.  
No. of unique values: 109682  
No. of duplicates: 1548  

In [59]:
column = dataset["Name"]
column

0                                                    PRAYAS
1                               PONDICHERRYWOMENSCONFERENCE
2                                  SHABRI SAMAJ SEWA SAMITI
3                         ANAND GANGA SAMAJIK SIKSHA SAMITI
4                               Himaliyan Gram Vikas Samiti
                                ...                        
111924                            Hariom Samaj Vikas Samiti
111925    narmadanchal naya jeevan jan kalyaan seva sami...
111926                    Mathura Prasad Gramodyog Sansthan
111927                   Shree Swaminarayan Education Trust
111928                                      Srijan Sansthan
Name: Name, Length: 111929, dtype: object

In [64]:
# Number of empty strings/missing values
print("Invalid:", sum(column == " ") + sum(column.isnull()))
print("No. of unique values:", len(column.unique()))
# Check for duplicates
counter = dict(Counter(column))
duplicates = { key:[value] for key, value in counter.items() if value > 1}
print("No. of duplicates:", len(duplicates))
table = pd.DataFrame.from_dict(duplicates)
table = table.melt(var_name="Duplicate Names", value_name="Count").sort_values(by=["Count"], ascending=False).reset_index(drop=True)
table

Invalid: 86142
No. of unique values: 24253
No. of duplicates: 202


Duplicate Names  Count
0                                                  NaN  86142
1                                              http://    859
2    https://ngodarpan.gov.in/index.php/ngo/primaryngo    159
3                             https://ngodarpan.gov.in     57
4     http://ngodarpan.gov.in/index.php/ngo/primaryngo     56
..                                                 ...    ...
197                        http://www.saraswatigoi.com      2
198                               http://www.msmhc.org      2
199                     http://www.poddarinstitute.org      2
200                         http://www.svnycindia.org/      2
201                        http://www.umeedsociety.com      2

[202 rows x 2 columns]

In [56]:
# Example
dataset[column == table.iloc[0,0]].head()

Name ngo url        Mobile         UniqueID Off phone1  \
23563  CATHOLIC CHURCH     NaN  9.427830e+09  GJ/2017/0168939        NaN   
26416  CATHOLIC CHURCH     NaN  9.998133e+09  GJ/2017/0172648        NaN   
26884  CATHOLIC CHURCH     NaN  9.426880e+09  GJ/2017/0173225        NaN   
27081  CATHOLIC CHURCH     NaN  9.426513e+09  GJ/2017/0173470        NaN   
27999  CATHOLIC CHURCH     NaN  9.426389e+09  GJ/2017/0174578        NaN   

                                Email  \
23563   CATHOLICCHURCHVYARA@GMAIL.COM   
26416              ccvadtal@gmail.com   
26884  catholicchurchzaroli@gmail.com   
27081               frsjraj@gmail.com   
27999             frarulhmt@gmail.com   

                                       Major Activities1  \
23563                                  PRIEST MANTENANCE   
26416                               RELIGIOUS AND SOCIAL   
26884  THE OBJECT OF THE TRUST ARE RELIGIOUS EDUCATIO...   
27081  THE OBJECT OF THE TRUST ARE RELIGIOUS AND CHAR...   
27999  Trust has mainly involved in village animation...   

      operational states db  \
23563              GUJARAT,   
26416              GUJARAT,   
26884              GUJARAT,   
27081              GUJARAT,   
27999              GUJARAT,   

                                       issues working db  \
23563  Children,Education & Literacy,Health & Family ...   
26416                    Education & Literacy,Any Other,   
26884                    Education & Literacy,Any Other,   
27081                    Education & Literacy,Any Other,   
27999                              Education & Literacy,   

      operational district db  ... Asisstant Secretary mobile  \
23563          GUJARAT->Tapi,  ...                        NaN   
26416        GUJARAT->Anand ,  ...                        NaN   
26884        GUJARAT->Valsad,  ...                        NaN   
27081        GUJARAT->Anand ,  ...                        NaN   
27999       GUJARAT->Arvalli,  ...                        NaN   

      Board Member name Board Member email Board Member mobile  \
23563               NaN                NaN                 NaN   
26416               NaN                NaN                 NaN   
26884               NaN                NaN                 NaN   
27081               NaN                NaN                 NaN   
27999               NaN                NaN                 NaN   

      Vice Chairman name Vice Chairman email Vice Chairman mobile  \
23563                NaN                 NaN                  NaN   
26416                NaN                 NaN                  NaN   
26884                NaN                 NaN                  NaN   
27081                NaN                 NaN                  NaN   
27999                NaN                 NaN                  NaN   

                           Member name               Member email  \
23563                              NaN                        NaN   
26416                      SELVIN CRUZ         ccvadtal@gmail.com   
26884        VASAVA KANTILAL HIMATSING  kantivasava1959@gmail.com   
27081  ARASAKUMAR DEVASAGAYAM RAYAPPAN        arayappan@gmail.com   
27999            Kamjibhai Nemaji Dund        KAMJIDUND@GMAIL.COM   

      Member mobile  
23563           NaN  
26416  9.763823e+09  
26884  9.904656e+09  
27081  9.426513e+09  
27999  9.979210e+09  

[5 rows x 42 columns]

Same ```name``` does not mean duplicate rows.

### ngo url
<a class="anchor" id="4.2"></a>
URL for NGO.  
No. of unique values: 24253  
No. of duplicates values: 202   
A lot of NGOs were confused in the information filling process, and pasted the NGO darpan URL instead of the URL to their NGO's website if there is one: the first 13. So there are a large number of invalid URLs.  
Additionally, a large number of urls cannot be reached.

In [62]:
column = dataset["ngo url"]
column

0                                  NaN
1                                  NaN
2            http://ssssamitibhind.org
3                                  NaN
4                                  NaN
                      ...             
111924                         http://
111925                             NaN
111926         http://mathuravikas1977
111927                             NaN
111928    http:/www.srijansansthan.com
Name: ngo url, Length: 111929, dtype: object

In [65]:
# Number of empty strings/missing values
print("Invalid:", sum(column == " ") + sum(column.isnull()))

print("No. of unique values:", len(column.unique()))

# Check for duplicates
counter = dict(Counter(column))
duplicates = { key:[value] for key, value in counter.items() if value > 1}
print("No. of Duplicates:", len(duplicates))

table = pd.DataFrame.from_dict(duplicates)
table = table.melt(var_name="Duplicate URLs", value_name="Count").sort_values(by=["Count"], ascending=False).reset_index(drop=True)
table

Invalid: 86142
No. of unique values: 24253
No. of Duplicates: 202


Duplicate URLs  Count
0                                                  NaN  86142
1                                              http://    859
2    https://ngodarpan.gov.in/index.php/ngo/primaryngo    159
3                             https://ngodarpan.gov.in     57
4     http://ngodarpan.gov.in/index.php/ngo/primaryngo     56
..                                                 ...    ...
197                        http://www.saraswatigoi.com      2
198                               http://www.msmhc.org      2
199                     http://www.poddarinstitute.org      2
200                         http://www.svnycindia.org/      2
201                        http://www.umeedsociety.com      2

[202 rows x 2 columns]

A lot of NGOs were confused in the information filling process, and pasted the NGO darpan URL instead of the URL to their NGO's website if there is one: the first 13. So there are a large number of invalid URLs. Additionally, a large number of urls cannot be reached.

In [70]:
table.iloc[13:]

Duplicate URLs  Count
13            http://www.jss.nic.in      8
14        http://www.srigurudev.org      8
15            http://www.durbar.org      7
16                http://jss.nic.in      7
17      http://www.ngo.india.gov.in      6
..                              ...    ...
197     http://www.saraswatigoi.com      2
198            http://www.msmhc.org      2
199  http://www.poddarinstitute.org      2
200      http://www.svnycindia.org/      2
201     http://www.umeedsociety.com      2

[189 rows x 2 columns]

In [71]:
# Example
dataset[column == table.iloc[13,0]].head()

Name                ngo url        Mobile  \
1532  STATE RESOURCE CENTER DISPUR ASSAM  http://www.jss.nic.in  9.435328e+09   
1603        jan shikshan sansthan raigad  http://www.jss.nic.in  9.922999e+10   
1611    JAN SHIKSHAN SANSHTHAN CHANDAULI  http://www.jss.nic.in  8.924071e+09   
1660          JAN SHIKSHAN SANSTHAN GUNA  http://www.jss.nic.in  9.425633e+09   
1702         JAN SHIKSHAN SANSTHAN BASTI  http://www.jss.nic.in  9.415174e+09   

             UniqueID    Off phone1                    Email  \
1532  AS/2016/0104590  0361-2388990      srcdispur@yahoo.com   
1603  MH/2016/0104734  02141-227932         jssrgd@gmail.com   
1611  UP/2016/0104747  05412-260625  jsschandauli1@gmail.com   
1660  MP/2016/0104852  07542-252375        jssguna@gmail.com   
1702  UP/2016/0104945  05542-209210      jssbasti44@yahoo.in   

                                      Major Activities1 operational states db  \
1532  The critical focus of SRC Dispur is giving tec...                ASSAM,   
1603  14. Jan Shikshan Sansthan Raigad ,Maharashtra ...          MAHARASHTRA,   
1611  JAN SHIKSHAN SANSHTHAN IS ORGNISING VOCATIONAL...        UTTAR PRADESH,   
1660  Jan Shikshan Sansthan is working of Vocational...       MADHYA PRADESH,   
1702  We are conducting the activities by govt of In...        UTTAR PRADESH,   

                                      issues working db  \
1532  Education & Literacy,Human Rights,Vocational T...   
1603                               Vocational Training,   
1611                               Vocational Training,   
1660  Civic Issues,Drinking Water,Education & Litera...   
1702                              Education & Literacy,   

                            operational district db  ...  \
1532  ASSAM->Barpeta, Dhubri, Hailakandi, Tinsukia,  ...   
1603                          MAHARASHTRA->Raigarh,  ...   
1611           UTTAR PRADESH->Chandauli, Chandauli,  ...   
1660                          MADHYA PRADESH->Guna,  ...   
1702                          UTTAR PRADESH->Basti,  ...   

     Asisstant Secretary mobile Board Member name Board Member email  \
1532                        NaN               NaN                NaN   
1603                        NaN              Advt   jssrgd@gmail.com   
1611                        NaN               NaN                NaN   
1660                        NaN               NaN                NaN   
1702                        NaN               NaN                NaN   

     Board Member mobile Vice Chairman name    Vice Chairman email  \
1532                 NaN                NaN                    NaN   
1603        9.372212e+09                Mrs       jssrgd@gmail.com   
1611                 NaN                NaN                    NaN   
1660                 NaN      Vishnu Sharma  yuvacollege@ymail.com   
1702                 NaN                NaN                    NaN   

     Vice Chairman mobile     Member name           Member email Member mobile  
1532                  NaN             NaN                    NaN           NaN  
1603         9.766168e+09             NaN                    NaN           NaN  
1611                  NaN             NaN                    NaN           NaN  
1660         9.893453e+09  Manisha Pandey  parthpandey@gmail.com  9.926540e+09  
1702                  NaN             NaN                    NaN           NaN  

[5 rows x 42 columns]

Duplicates for ```ngo url``` do not mean duplicate rows.

### Mobile
<a class="anchor" id="4.3"></a>
Mobile number. 
Incorrect dtype.
No. of unique values: 24253  
No. of duplicates values: 202   

In [77]:
column = dataset["Mobile"]
column

0         9.778080e+09
1         9.443253e+09
2         7.828394e+09
3         9.450678e+09
4         9.412037e+09
              ...     
111924    9.335113e+09
111925    9.425042e+09
111926    9.452122e+09
111927    9.727767e+09
111928    9.667617e+09
Name: Mobile, Length: 111929, dtype: float64